bitshuffle/library.py

```diff
@@ -0,0 +1,231 @@
+        return buf.getvalue()
+
+    def gzip_decompress(data):
+        # pylint: disable=missing-docstring
+        with gzip.GzipFile(fileobj=io.BytesIO(data)) as f:
+            return f.read()
+
+    gzip.compress = gzip_compress
+    gzip.decompress = gzip_decompress
+    del gzip_compress, gzip_decompress
+
+
+def warn(integer, *argv, **kwargs):
+    '''Issue a warning to stderr for every argument given
+    Prepends colon to every successive argument
+    kwargs are ignored except for severity'''
+    if 'severity' not in kwargs.keys():
+        kwargs['severity'] = 2
+    error = "%s %d: %s" % (LEVELS[kwargs['severity']],
+                           integer, ERRORS[integer])
+    for arg in argv:
+        error += ": " + arg
+    stderr.write(error + '\n')
+
+
+def exit_with_error(integer, *argv, **kwargs):
+    '''Issue a warning and exit with return code `integer`.
+    Each successive argument is added to the warning, prepended by a colon.
+    kwargs are ignored, excluding severity'''
+    if 'severity' not in kwargs.keys():
+        kwargs['severity'] = 3
+    warn(integer, *argv, severity=kwargs['severity'])
+    exit_with_code(integer)
+
+
+def exit_successfully():
+    '''exit, successfully'''
+    if DEBUG:
+        warn(0, 0)
+    exit_with_code(0)
+
+
+def shasum(data):
+    '''Return sha256 sum of input, encdoded in base16'''
+    return sha256(data).hexdigest()
+
+
+def encode_data(data, chunksize=2048, compresslevel=5, compress=bz2.compress):
+    """encode_data
+
+    Compress the given data (which should be bytes), chunk it into chunksize
+    sized chunks, base64 encode the chunks, and return the lot as a list of
+    chunks, which are strings.
+
+    :param data: bytes
+    :param compresstype: string: bz2 or gzip
+    :param compresslevel: int: 1-9 inclusive
+    """
+
+    data = compress(data, compresslevel=compresslevel)
+    chunks = []
+    chunkptr = 0
+    while True:
+        chunk = data[chunkptr:chunkptr + chunksize]
+        chunkptr += chunksize
+
+        chunks.append(base64.b64encode(chunk))
+
+        if chunkptr >= len(data):
+            chunk = data[chunkptr:]
+            chunks.append(base64.b64encode(chunk))
+            break
+
+    return [chunk for chunk in chunks if len(chunk) > 0]
+
+
+def encode_packet(data, seqnum, seqmax, compress=bz2.compress, msg=DEFAULT_MSG,
+                  file_hash=None):
+    """encode_packet
+
+    Take an already encoded data string and encode it to a BitShuffle data
+    packet.
+
+    :param data: bytes
+    """
+
+    encoding = "base64"
+    packet_hash = shasum(data)
+    data = data.decode()
+
+    fmt = "((<<{}|{}|{}|{}|{}|{}|{}|{}"
+    packet = fmt.format(msg, COMPATLEVEL, encoding, compress.__module__,
+                        seqnum, seqmax, packet_hash, data)
+    if file_hash is not None:
+        packet += '|'
+        packet += file_hash
+    packet += '>>))'
+    return packet
+
+
+def encode_file(fhandle=stdin, chunksize=2048, compresslevel=5,
```

Comment:
It is not necessary to deal with this for #60, however at some point, we should check if fhandle is `bytes` or `bytearray`, in which case we should use it as data rather than opening it as a handle.

Reply:
added https://github.com/charlesdaniels/bitshuffle/issues/61